File: linked_list_problems/deleteNode.cpp
/**
 *  Problem: Only given a pointer to node of the linked list, delete the node.
 */
#include <iostream>

struct Node {
  int data;
  Node * next;
  Node( int d ) : data{ d }, next{ nullptr } { }
};

void insert( Node * & head, int data )
{
  Node *newNode = new Node(data);
  if ( head == nullptr ) {
    head = newNode;
  } else {
    Node * temp = head;
    while ( temp->next != nullptr ) {
      temp = temp->next;
    }
    temp->next = newNode;
  }
}

void printList( Node * head )
{
  while( head ) {
    std::cout << head->data << "-->";
    head = head->next;
  }
  std::cout << "NULL" << std::endl;
}

void deleteNode( Node * node )
{
  // return if node is null
  if ( node == nullptr ) {
    return;
  }

  //this method won't work if we are given last node
  if ( node->next == nullptr ) {
    return;
  }

  Node * nextNode = node->next;
  node->data = nextNode->data;
  node->next = nextNode->next;
  delete nextNode;
}

//Delete a Linked List node at a given position
void deleteNode(Node **head_ref, int position)
{
     
    // If linked list is empty
    if (*head_ref == NULL)
        return;
     
    // Store head node
    Node* temp = *head_ref;
 
    // If head needs to be removed
    if (position == 0)
    {
         
        // Change head
        *head_ref = temp->next;
         
        // Free old head
        free(temp);            
        return;
    }
 
    // Find previous node of the node to be deleted
    for(int i = 0; temp != NULL && i < position - 1; i++)
        temp = temp->next;
 
    // If position is more than number of nodes
    if (temp == NULL || temp->next == NULL)
        return;
 
    // Node temp->next is the node to be deleted
    // Store pointer to the next of node to be deleted
     Node *next = temp->next->next;
 
    // Unlink the node from linked list
    free(temp->next); // Free memory
     
     
    temp->next = next;
}
 



int main()
{
  Node * head = nullptr;
  insert( head, 1 );
  insert( head, 12 );
  insert( head, 2 );
  insert( head, 3 );
  insert( head, 4 );
  printList( head );
  std::cout << "Deleting node with data 12 \n";
  deleteNode( head->next );
  printList( head );
  return 0;
}

File: linked_list_problems/delete_node.py
# Problem: Only given a node of the linked list, delete the node.


class Node:
    """Representation of a node of a linked list"""
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    """Representation of a linked list"""
    def __init__(self):
        self.head = None

    def push(self, data):
        """Pushes a new node to the linkedlist, the new node becomes head
        Params:
            data of the new node.
        """
        new_node = Node(data)
        new_node.next = self.head
        self.head = new_node

    def print_list(self):
        """Prints the linked list"""
        itr = self.head
        while itr:
            if itr.next is None:
                print(itr.data)
            else:
                print(str(itr.data)+ "-->", end=" ")
            itr = itr.next

    def delete_node(self, node):
        """Deletes a given node from the linked list
        Params:
            node: node to be deleted
        """
        itr = self.head
        while itr.next != node:
            itr = itr.next
        if itr is not None:
            itr.next = node.next

        node.next = None

    def get_head(self):
        return self.head


if __name__ == '__main__':
    linkedList = LinkedList()
    linkedList.push(1)
    linkedList.push(2)
    linkedList.push(3)
    linkedList.push(4)
    linkedList.push(4)
    print("List before deleting:")
    linkedList.print_list()

    # Delete second node from the list.
    linkedList.delete_node(linkedList.get_head().next)
    print("List after deleting:")
    linkedList.print_list() 